### PR TITLE
Add export tasks

### PR DESCRIPTION
**Pull request by xianbaoqian**
Context: https://huggingface.slack.com/archives/C02ERUY9TKK/p1669743642830359

### CODE REVIEW

**Comment by xianbaoqian** [comment]
image

**Comment by xianbaoqian** [comment]
{
    "adapter_transformers": [
        "question-answering",
        "text-classification",
        "token-classification"
    ],
    "allennlp": [
        "question-answering"
    ],
    "asteroid": [
        "audio-source-separation",
        "audio-to-audio"
    ],
    "diffusers": [
        "text-to-image"
    ],
    "doctr": [
        "object-detection"
    ],
    "espnet": [
        "text-to-speech",
        "automatic-speech-recognition"
    ],
    "fairseq": [
        "text-to-speech",
        "audio-to-audio"
    ],
    "fastai": [
        "image-classification"
    ],
    "fasttext": [
        "feature-extraction",
        "text-classification"
    ],
    "flair": [
        "token-classification"
    ],
    "k2_sherpa": [
        "automatic-speech-recognition"
    ],
    "keras": [
        "image-classification"
    ],
    "nemo": [
        "automatic-speech-recognition"
    ],
    "paddlenlp": [
        "conversational",
        "fill-mask"
    ],
    "pyannote_audio": [
        "automatic-speech-recognition"
    ],
    "sentence_transformers": [
        "feature-extraction",
        "sentence-similarity"
    ],
    "sklearn": [
        "tabular-classification",
        "tabular-regression",
        "text-classification"
    ],
    "spacy": [
        "token-classification",
        "text-classification",
        "sentence-similarity"
    ],
    "speechbrain": [
        "audio-classification",
        "audio-to-audio",
        "automatic-speech-recognition",
        "text-to-speech",
        "text2text-generation"
    ],
    "stanza": [
        "token-classification"
    ],
    "superb": [
        "automatic-speech-recognition",
        "speech-segmentation"
    ],
    "timm": [
        "image-classification"
    ]
}

File: .github/workflows/python-api-export-tasks.yaml [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+      - uses: actions/checkout@v2
+
+      - name: Set up Python ${{ matrix.python-version }}
+        uses: actions/setup-python@v2
+        with:
+          python-version: ${{ matrix.python-version }}
+
+      - run: python scripts/export_tasks.py
+        env: # Or as an environment variable
+          API_TOKEN: ${{ secrets.USER_API_TOKEN }}
```
Comment (osanseviero) [inline .github/workflows/python-api-export-tasks.yaml:26]: Is this needed?

File: scripts/export_tasks.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
+import ast
```
Comment (osanseviero) [inline scripts/export_tasks.py:1]: Maybe add a short docstring at the top explaining what this does/what's the output of the script

Comment (osanseviero) [reply scripts/export_tasks.py:1]: Also mention the limitation that it does not include transformers

File: scripts/export_tasks.py [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+import os
+import pathlib
+import json
+
+lib_to_task_map = {}
+
+
+def extract_tasks(library_name, variable_name, value):
```
Comment (osanseviero) [inline scripts/export_tasks.py:9]: Nit, a good way to indicate this method is intended to be private

```suggestion
def _extract_tasks(library_name, variable_name, value):
```

Comment (osanseviero) [inline scripts/export_tasks.py:9]: Would be great if you could add docstrings in these methods. What is `variable_name` and what is `value`? Would also be good to add some typings

File: scripts/export_tasks.py [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    if variable_name == "ALLOWED_TASKS":
+        if isinstance(value, ast.Dict):
+            for key in value.keys:
+                lib_to_task_map.setdefault(library_name, []).append(key.value)
```
Comment (osanseviero) [inline scripts/export_tasks.py:13]: Rather than using `setdefault`, a nice way is to use `defaultdict` when you define `lib_to_task_map. Then you can specify a default value

```python
from collections import defaultdict
dct = defaultdict(list)
```

File: scripts/export_tasks.py [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+
+
+def traverse_global_assignments(library_name, file_content, handler):
```
Comment (osanseviero) [inline scripts/export_tasks.py:16]: Same as above, some docstrings here would be useful

File: scripts/export_tasks.py [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+    for element in ast.parse(file_content).body:
+        # Typical case
+        if isinstance(element, ast.AnnAssign):
+            handler(library_name, element.target.id, element.value)
```
Comment (osanseviero) [inline scripts/export_tasks.py:20]: Out of curiosity, what is `target.id` here?

File: .github/workflows/python-api-export-tasks.yaml [hunk 7]
```diff
@@ -0,0 +1,26 @@
+name: Export tasks
+on:
+  pull_request:
+    paths:
+      - "**"
```
Comment (osanseviero) [inline .github/workflows/python-api-export-tasks.yaml:5]: I would likely run this on merge to main rather than for every pull request

**Review by osanseviero — COMMENTED** [review]
Very nice! The solution seems fine to me. I left some suggestions to make this easier to maintain and understand, but overall LG

File: scripts/export_tasks.py (hunk 6 above)
Comment (xianbaoqian) [reply scripts/export_tasks.py:20]: Sure. target is name. it contains both id which is the variable name and also a context to differentiate loading, deleting and assignment. Ref: https://docs.python.org/3/library/ast.html#ast.Store

**Review by osanseviero — APPROVED** [review]
LGTM!
